enumerables.rb

```diff
@@ -0,0 +1,196 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    if is_a? Array
+      i = 0
+      while i < size
+        yield(self[i])
+        i += 1
+      end
+    elsif is_a? Hash
+      while i < size
+        yield(values[i], keys[i])
+        i += 1
+      end
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    arr_to_work = self
+    if arr_to_work.is_a? Array
+      i = 0
+      while i < arr_to_work.length
+        yield(arr_to_work[i], i)
+        i += 1
+      end
+    elsif arr_to_work.is_a? Hash
+      j = 0
+      while j < arr_to_work.length
+        yield([arr_to_work.keys[j], arr_to_work.values[j]], j)
+        j += 1
+      end
+    end
+    self
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    arr_to_work = self
+    if is_a? Array
+      myarr = []
+      arr_to_work.my_each do |item|
+        myarr.push(item) if yield(item)
+      end
+    elsif is_a? Hash
+      myarr = {}
+      arr_to_work.my_each do |k, v|
+        myarr[k] = v if yield(k, v)
+      end
+    end
+    myarr
+  end
+
+  def my_all?(pattern = nil) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+    return true unless block_given? || !pattern.nil?
+
+    arr_to_work = self
+    if pattern
+      arr_to_work.my_each do |item|
+        return false unless item.is_a? pattern
+      end
+    elsif is_a? Array
+      arr_to_work.my_each do |item|
+        return false unless yield(item)
+      end
+    elsif is_a? Hash
+      arr_to_work.my_each do |k, v|
+        return false unless yield(k, v)
+      end
+    end
+    true
+  end
+
+  def my_any?(pattern = nil) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+    return false unless block_given? || !pattern.nil?
+
+    arr_to_work = self
+    if pattern
+      arr_to_work.my_each do |item|
+        return true if item.is_a? pattern
+      end
+
+    elsif is_a? Array
+      arr_to_work.my_each do |item|
+        return true if yield(item)
+      end
+    elsif is_a? Hash
+      arr_to_work.my_each do |k, v|
+        return true if yield(k, v)
+      end
+    end
+    false
+  end
```

Comment:
1. When no block or argument is given returns true if at least one of the collection is not false or nil

`[false, nil, 4].my_any?` should return true
`[false, nil].my_any?` should return false

2.  When a Regex is passed as an argument returns false if none of the collection matches the Regex

`['ant', 'bear', 'boy'].my_any?(/t/)` shoud returns true
`['car', 'bear', 'boy'].my_any?(/t/)` shoud returns false

3. When a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

`p [1,2,3,4].my_any?(4)` should returns true
`p [1,2,3,5].my_any?(4)` should returns false